File: scala_api/src/main/scala/application/dataAnalysis/ml/regression/Regressors.scala
package application.dataAnalysis.ml.regression

import application.dataAnalysis.interfaces.{ModelCreateProcess, ModelPrepare}
import application.dataAnalysis.ml.ClassificatorsAndRegressors
import application.models.Hyperparameters
import application.models.SummaryOfMerics.{RegressionSummaryResult, SummaryResult}
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.util.MLWritable
import org.apache.spark.ml.{PipelineModel, PipelineStage}
import org.apache.spark.sql.{DataFrame, Dataset, Row}

case class Regressors(
 override val data: DataFrame,
 override val train: Dataset[Row],
 override val test: Dataset[Row],
 override val maxCategories: Option[Int] = None
) extends ClassificatorsAndRegressors(data, train, test, maxCategories) with ModelCreateProcess {

  override def makeFitAndTransform(regressionModel: PipelineStage with MLWritable): DataFrame = {
    pipeline.setStages(Array(featureIndexer, regressionModel))
    val model: PipelineModel = pipeline.fit(train)
    val predictions: DataFrame = model.transform(test)
    safeResources(model)
    predictions
  }

  override def chooseAndCreateModel(hyperparameters: Hyperparameters, methodName: String): ModelPrepare = {
    methodName match {
      case "Linear regression" => hyperparameters.linearRegressionModel.getOrElse(LinearRegressionPattern())
      case "Generalized Linear regression" => hyperparameters.generalizedLinearRegressionModel.getOrElse(GeneralizedLinearRegressionPattern())
      case "Decision tree regression" => hyperparameters.decisionTreeRegModel.getOrElse(DecisionTreeRegPattern())
      case "Random forest regression" => hyperparameters.randomForrestRegModel.getOrElse(RandomForrestRegPattern())
      case _ => hyperparameters.linearRegressionModel.getOrElse(LinearRegressionPattern())
    }
  }

  override def calculateMetrics(predictions: DataFrame): SummaryResult =  {
    val regressionEvaluatorRmse = new RegressionEvaluator
    val regressionEvaluatorR2 = new RegressionEvaluator
    val evaluatorRMSE = regressionEvaluatorRmse
      .setPredictionCol("prediction")
      .setMetricName("rmse")
    val evaluatorR2 = regressionEvaluatorR2
      .setPredictionCol("prediction")
      .setMetricName("r2")
    val rmse = evaluatorRMSE.evaluate(predictions)
    val r2 = evaluatorR2.evaluate(predictions)
    val metrics = RegressionSummaryResult(r2, rmse)
    SummaryResult(regressionMetrics = Some(metrics))
  }

}

object Regressors{
  val indexOfFeatureIndexer: Int = 0
  val predictionColumn = "prediction"
}

File: python_api/SparkVersion/ml/regression/Regressors.py
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.feature import VectorIndexer

import Context
from Models.RegressionSummaryResult import RegressionSummaryResult
from Models.SummaryResult import SummaryResult
from SparkVersion.ml.regression.DecisionTreeRegressionModel import DecisionTreeRegressionModel
from SparkVersion.ml.regression.GeneralizedLinearRegressionModel import GeneralizedLinearRegressionModel
from SparkVersion.ml.regression.LinearRegressionModel import LinearRegressionModel
from Context import getPipelineModelFilePath
from SparkVersion.ml.regression.RandomForrestRegressionModel import RandomForrestRegressionModel


class Regressors:
    def __init__(self, data, train, test, maxCategories=4):
        self.data = data
        self.train = train
        self.test = test
        self.maxCategories = maxCategories

    @staticmethod
    def chooseMethod(method, hyperparameters):
        if method == "Linear regression":
            return LinearRegressionModel(hyperparameters).prepareModel()
        elif method == "Generalized Linear regression":
            return GeneralizedLinearRegressionModel(hyperparameters).prepareModel()
        elif method == "Decision tree regression":
            return DecisionTreeRegressionModel(hyperparameters).prepareModel()
        else:
            return RandomForrestRegressionModel(hyperparameters).prepareModel()

    def fitAndTransform(self, regressionModel):
        pipeline = Pipeline().setStages([self.featureIndexer(), regressionModel])
        model = pipeline.fit(self.train)
        predictions = model.transform(self.test)
        if Context.saveModels:
            model.write().overwrite().save(getPipelineModelFilePath())
        return predictions

    @staticmethod
    def calculateMetrics(predictions):
        evaluatorRMSE = RegressionEvaluator(
            predictionCol="prediction",
            metricName="rmse"
        )
        evaluatorR2 = RegressionEvaluator(
          predictionCol="prediction",
          metricName ="r2"
        )
        rmse = evaluatorRMSE.evaluate(predictions)
        r2 = evaluatorR2.evaluate(predictions)
        metrics = RegressionSummaryResult(r2, rmse)
        return SummaryResult(regressionMetrics = metrics)

    def featureIndexer(self):
        return VectorIndexer(
            inputCol="features",
            outputCol="indexedFeatures",
            maxCategories=self.maxCategories
        ).fit(self.data)

    @staticmethod
    def indexOfFeatureIndexer():
        return 0

    @staticmethod
    def predictionColumn():
        return "prediction"
